### PR TITLE
packets_creation: add the icub packet creation via gh action

### PR DESCRIPTION
**Pull request by Nicogene**
This PR adds the creation of the `icub` packet via gh action

See https://github.com/Nicogene/icub-main/actions/runs/463171210 for reference and here the deb uploaded

immagine

### CODE REVIEW

File: .github/workflows/packages_creation.yml [hunk 1]
```diff
@@ -72,6 +72,8 @@ jobs:
         sudo apt-get update
         sudo apt-get install -y git
         sudo apt-get install -y lsb-release
+        sudo apt-get install -y build-essential
+        sudo apt-get install -y qml-module-qt-labs-folderlistmodel qml-module-qt-labs-settings libopenni2-dev libi2c-dev libv4l-dev libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-libav gstreamer1.0-tools libgraphviz-dev libqcustomplot-dev
```
Comment (pattacini) [inline .github/workflows/packages_creation.yml:76]: `apt` might be a better choice than `apt-get`

File: .github/workflows/packages_creation.yml [hunk 2]
```diff
@@ -97,14 +108,19 @@ jobs:
...
-        asset_path: ${{ steps.package_generation.outputs.ICUB_PACKAGE_NAME }}
+        asset_path: ${{ steps.package_generation_icub-common.outputs.ICUB_COMMON_PACKAGE_NAME }}
         asset_content_type: application/vnd.debian.binary-package
 
+    - name: Upload Release Asset icub-main
+      uses: shogo82148/actions-upload-release-asset@v1
+      with:
+        upload_url: ${{ steps.read_upload_url.outputs.UPLOAD_URL }}
+        asset_path: ${{ steps.package_generation_icub-main.outputs.ICUB_MAIN_PACKAGE_NAME }}
+        asset_content_type: application/vnd.debian.binary-package
```
Comment (pattacini) [inline .github/workflows/packages_creation.yml:123]: Please, let's use more standard actions as `actions/upload-release-asset`.

Take a look at https://github.com/icub-tech-iit/ventilator-FI5/blob/master/.github/workflows/create-release.yml for an example.

File: .github/workflows/packages_creation.yml [hunk 3]
```diff
@@ -127,6 +142,8 @@ jobs:
         apt-get update
         apt-get install -y git
         apt-get install -y lsb-release
+        apt-get install -y build-essential wget
+        apt-get install -y libsqlite3-dev qml-module-qt-labs-folderlistmodel qml-module-qt-labs-settings libopenni2-dev libi2c-dev libv4l-dev libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-libav gstreamer1.0-tools libgraphviz-dev libqcustomplot-dev
```
Comment (pattacini) [inline .github/workflows/packages_creation.yml:146]: stick to `apt`.

File: .github/workflows/packages_creation.yml [hunk 4]
```diff
@@ -152,12 +177,17 @@ jobs:
       shell: bash
       run: echo ::set-output name=UPLOAD_URL::$( cat UPLOAD_URL.txt )
 
-    - name: Upload Release Asset
-      id: upload-release-asset
+    - name: Upload Release asset icub-common
+      uses: shogo82148/actions-upload-release-asset@v1
```
Comment (pattacini) [inline .github/workflows/packages_creation.yml:181]: Same comment as above.

**Review by pattacini — CHANGES_REQUESTED** [review]
Nice addition @Nicogene 👍🏻 

I've got a few points to address.

**Comment by pattacini** [comment]
Another problem I can see is that we tend to modify the release tag a few times before carving it on the stone.
Therefore, I'd avoid triggering the action upon the creation of just the tag.

Instead, I'd go for one of the following options:
1. Trigger upon creation of the release.
2. Make the job manual using the `workflow_dispatch` event.

**Comment by traversaro** [comment]
> Another problem I can see is that we tend to modify the release tag a few times before carving it on the stone to fix some last minutes mistakes.

No directly related to this PR, but I guess this is an residual related to when releases were "expensive". If the release process is completely automated, there is nothing particularly problematic in quickly releasing small patch releases that fix small bugs or problems.

> Instead, I'd go for one of the following options:
> 
>     1. Trigger upon manual creation of the release.
> 
>     2. Make the job manual using the `workflow_dispatch` event.
> 
> 
> cc @traversaro @Nicogene

The problem of using `workflow_dispatch` is that then you are executing the code on a given branch, that may be different from the code that was tagged in the release. The `release` event offers a lot of activity types (see https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#release) that I think can be used in this case.

File: .ci/packages_vars.sh [hunk 5]
```diff
@@ -9,16 +9,47 @@
...
 SUPPORTED_DISTRO_LIST="buster bionic focal"
 SUPPORTED_TARGET_LIST="amd64"
 CMAKE_MIN_REQ_VER="3.12.0"
 ICUB_DEPS_COMMON="libace-dev libboost-filesystem-dev libboost-system-dev libboost-thread-dev libc6 python3 libgsl0-dev libncurses5-dev libsdl1.2-dev subversion git gfortran libxmu-dev libode-dev wget unzip qtbase5-dev qtdeclarative5-dev qtmultimedia5-dev libqt5svg5 libqt5opengl5-dev libopencv-dev freeglut3-dev libtinyxml-dev libblas-dev coinor-libipopt-dev liblapack-dev libmumps-dev qml-module-qtmultimedia qml-module-qtquick-dialogs qml-module-qtquick-controls libedit-dev libeigen3-dev libjpeg-dev libsimbody-dev libxml2-dev libjs-underscore portaudio19-dev ${YCM_PACKAGE}"
 ICUB_DEPS_bionic="libode6"
 ICUB_DEPS_focal="libode8"
 ICUB_DEPS_buster="libode8"
-ICUB_REPO_URL="https://github.com/robotology/icub-main"
+#TO BE CHANGED!!!
+ICUB_REPO_URL="https://github.com/Nicogene/icub-main.git"
```
Comment (traversaro) [inline .ci/packages_vars.sh:32]: Leftover?

File: .ci/create_icub-main_deb.sh [hunk 6]
```diff
@@ -222,45 +228,175 @@ check_and_install_deps()
...
+
 }
 
 main()
 {
-  create_control_file
+  install_deps
+  build_icub
+  # Probably this step is not needed anymore after cmake components/modernization
+  fix_relocatable_files
```
Comment (traversaro) [inline .ci/create_icub-main_deb.sh:399]: I think it is still necessary for: 
* https://github.com/robotology/yarp/issues/2445
* https://github.com/robotology/yarp/issues/2444

File: .github/workflows/packages_creation.yml [hunk 7]
```diff
@@ -124,9 +139,11 @@ jobs:
...
       shell: bash
       run: |
-        apt-get update
-        apt-get install -y git
-        apt-get install -y lsb-release
+        apt update
+        apt install -y git
+        apt install -y lsb-release
+        apt install -y build-essential wget
+        apt install -y libsqlite3-dev qml-module-qt-labs-folderlistmodel qml-module-qt-labs-settings libopenni2-dev libi2c-dev libv4l-dev libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-libav gstreamer1.0-tools libgraphviz-dev libqcustomplot-dev
```
Comment (traversaro) [inline .github/workflows/packages_creation.yml:146]: Why is this necessary? The .deb package is happening in the chroot environment, and so the packages should be installed there, not in the host environment, right?

Comment (Nicogene) [reply .github/workflows/packages_creation.yml:146]: I am not using chroot, just using the host environment. If I am not missing something I think that using this infrastructure the chroot is not needed anymore, but correct me if I am wrong

**Comment by pattacini** [comment]
>No directly related to this PR, but I guess this is an residual related to when releases were "expensive". If the release process is completely automated, there is nothing particularly problematic in quickly releasing small patch releases that fix small bugs or problems.

I didn't explain exactly what I meant. In the recent past, @Nicogene can confirm that, we tagged the master and then craft the releases a few days afterward. In the time between, it (rarely) happened that the same tag was deleted and updated before announcing the "real" (corresponding) release. That's the point why I'm not really confident in triggering this workflow upon the creation of the tag but rather upon the creation of the actual release (and I'm pretty sure we can do that). 

Generally speaking, I concur with you @traversaro that once we release a piece of software we can then easily fix possible bugs with subsequent patches.

That said, option 1 (triggering upon releases rather than tag) is obviously preferable above option 2 - which by the way I meant to be using at the very same time when the release is ready.

File: .github/workflows/packages_creation.yml (hunk 7 above)
Comment (traversaro) [reply .github/workflows/packages_creation.yml:152]: If you are not using the `chroot`, and you are just running the job on `ubuntu-latest` (that I am not really sure if it is 18.04 or 20.04, but it is not relevant) how can you generate packages for different Ubuntu distributions?

Comment (traversaro) [reply .github/workflows/packages_creation.yml:152]: Sorry, I was confused, `preliminary-steps` run `ubuntu-latest`, then there are ubuntu-18.04 and ubuntu-20.04 jobs for `normal-build` and debian buster using `docker-build` . Probably it is unrelated this, but why can't we  build everything with `docker-build` and specifying the docker images for Ubuntu  18.04 and 20.04 ?

Comment (Nicogene) [reply .github/workflows/packages_creation.yml:152]: > can't we build everything with docker-build and specifying the docker images for Ubuntu 18.04 and 20.04?

Yes sure, I can do it! It would simplify the yml.
For now, I have a problem using the `upload-asset` action, the one I used worked, I don't understand why the standard one does not

Comment (traversaro) [reply .github/workflows/packages_creation.yml:152]: Perfect! For me if this version is working we can also merge it as it is, I just want to understand if there was a specific reason for this duplication, but we can cleanup it in the future.

**Comment by Nicogene** [comment]
Update

@pattacini now the action is triggered by the creation of a new release 👍🏻 
@traversaro the action uses now only docker, the yml is more slim 👍🏻 

See https://github.com/Nicogene/icub-main/actions/runs/469222425 for reference and the relative release https://github.com/Nicogene/icub-main/releases/tag/v1.0.57

**Review by pattacini — APPROVED** [review]
